gc.c

```diff
@@ -7018,7 +7048,7 @@ gc_mark_imemo(rb_objspace_t *objspace, VALUE obj)
         {
             const struct rb_callcache *cc = (const struct rb_callcache *)obj;
             // should not mark klass here
-            gc_mark(objspace, (VALUE)vm_cc_cme(cc));
+            rb_gc_mark_weak((VALUE *)&cc->cme_);
```

Comment:
Actually the `klass` field has weak reference semantics, too.  Currently, the `klass` field is cleared by the `obj_free` function for classes.  However, the following line in `gc_ref_update_imemo` (introduced in https://github.com/ruby/ruby/commit/087a2deccfb9d99961f1ce8526b80c5f72ee9a5d) may indicate that the `klass` filed may not always contain a valid object reference:

```c
// excerpt fro gc_ref_update_imemo
if (moved_or_living_object_strictly_p(objspace, cc->klass) &&
```

It may be a result that some `imemo:callcache` instance are not invalidated when `cc->klass` dies.

We may add `rb_gc_mark_weak((VALUE *)&cc->klass);` and see if it ever contains a reference to a dead class.

Reply:
Never mind the above comment. Ultimately I chose to mark the `klass` field as weak as well. I think that will fix the crashes on CI.

After 6dc15cc8895b28800d5c187929d846ac4eb7cd3f, during reference updating of compaction, we no longer directly set the klass of the CC as 0 when the CC has been invalidated. Therefore, we need mark it as weak so that it gets overwritten with a Qundef when the klass has been reclaimed.